xrpl/asyncio/wallet/wallet_generation.py

```diff
@@ -150,34 +154,40 @@ def process_faucet_host_url(input_url: str) -> str:
     return final_url
 
 
-def get_faucet_url(url: str, faucet_host: Optional[str] = None) -> str:
+def get_faucet_url(network_id: Optional[int], faucet_host: Optional[str] = None) -> str:
     """
-    Returns the URL of the faucet that should be used, based on whether the URL is from
-    a testnet or devnet client.
+    Returns the URL of the faucet that should be used, based on network_id and
+    faucet_host
 
     Args:
-        url: The URL that the client is using to access the ledger.
-        faucet_host: A custom host to use for funding a wallet.
+        network_id: The network_id corresponding to the XRPL network. This is parsed
+            from a ServerInfo() rippled response
+        faucet_host: A custom host to use for funding a wallet. This is useful because
+            network_id of sidechains might not be well-known. If faucet_url cannot be
+            gleaned from network_id alone, faucet_host is used
 
     Returns:
         The URL of the matching faucet.
 
     Raises:
-        XRPLFaucetException: if the provided URL is not for the testnet or devnet.
+        XRPLFaucetException: if the provided network_id does not correspond to testnet
+            or devnet.
     """
     if faucet_host is not None:
         return process_faucet_host_url(faucet_host)
-    if "altnet" in url or "testnet" in url:  # testnet
-        return _TEST_FAUCET_URL
-    if "sidechain-net2" in url:  # sidechain issuing chain devnet
-        raise XRPLFaucetException(
-            "Cannot fund an account on an issuing chain. Accounts must be created via "
-            "the bridge."
-        )
-    if "devnet" in url:  # devnet
-        return _DEV_FAUCET_URL
+
+    if network_id is not None:
+        if network_id == 1:
+            return _TEST_FAUCET_URL
+        elif network_id == 2:
+            return _DEV_FAUCET_URL
+        elif network_id == 0:
+            raise XRPLFaucetException("Cannot create faucet with a client on mainnet.")
+        elif network_id < 0:
+            raise XRPLFaucetException("network_id cannot be negative.")
+
     raise XRPLFaucetException(
-        "Cannot fund an account with a client that is not on the testnet or devnet."
+        "Cannot create faucet URL without network_id or the faucet_host information."
     )
```

Comment:
Thoughts on throwing this error in the main `generate_faucet_wallet` function instead of here? e.g.
```python
if network_id is None and faucet_host is None:
  raise XRPLFaucetException(...)
```

Reply:
I don't object to that idea, but how is that different from the current implementation?

At line 189 (the last line of the `get_faucet_url` method), we have established that `faucet_host == None`.  You are correct, an unknown devnet might have a networkID, but how can I determine the host_url from a networkID (say `1345` ?)

Reply:
> ```python
> if network_id is None and faucet_host is None:
>   raise XRPLFaucetException(...)
> ```
The said devnet will have a networkID, hence this condition needs to be:
```python
if faucet_host is None and networkID >= 1025:
  raise XRPLFaucetException(...)
```

The current implementation is equivalent to the above if-condition

Reply:
The difference comes from typing and failing fast. IMO it makes more sense for `get_faucet_url` to not accept `None` as a valid network ID.

You can't determine the host URL from an unknown Devnet, but the error message should be different.

Reply:
we cannot expect all networks to have a valid `NetworkID` isn't it? 
If we force `get_faucet_url` to accept valid networkIDs only, then some of the processing logic shifts into `generate_faucet_wallet` method.

```
def get_faucet_url(network_id: int) -> str:
    if network_id == 1:
        return _TEST_FAUCET_URL
    elif network_id == 2:
        return _DEV_FAUCET_URL
    elif network_id == 0:
        raise XRPLFaucetException("Cannot create faucet with a client on mainnet.")
    elif network_id < 0:
        raise XRPLFaucetException("network_id cannot be negative.")

    # unreachable code
    return ""

async def generate_faucet_wallet(
    client: Client,
    wallet: Optional[Wallet] = None,
    debug: bool = False,
    faucet_host: Optional[str] = None,
    usage_context: Optional[str] = None,
    user_agent: Optional[str] = "xrpl-py",
) -> Wallet:
    if not client.network_id:
        await get_network_id_and_build_version(client)

    if client.network_id is None and faucet_host is None:
        raise XRPLFaucetException(
            "Cannot create faucet URL without network_id or the faucet_host information"
        )

    if faucet_host is not None:
        faucet_url = process_faucet_host_url(faucet_host)
    else:
        faucet_url = get_faucet_url(client.network_id)
        
    # remaining code in the function ...

```

Reply:
That flow makes more sense to me, personally.

Reply:
alrighty

Reply:
updated sometime before [f06e5c1](https://github.com/XRPLF/xrpl-py/pull/715/commits/f06e5c15e618816ee7478909695f37a38618de1d)